xbmc/platform/linux/input/LibInputHandler.cpp

```diff
@@ -247,3 +172,55 @@ void CLibInputHandler::ProcessEvent(libinput_event *ev)
       break;
   }
 }
+
+void CLibInputHandler::DeviceAdded(libinput_device *dev)
+{
+  const char *sysname = libinput_device_get_sysname(dev);
+  const char *name = libinput_device_get_name(dev);
+
+  if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_TOUCH))
+  {
+    CLog::Log(LOGDEBUG, "CLibInputHandler::%s - touch type device added: %s (%s)", __FUNCTION__, name, sysname);
+    m_devices.push_back(libinput_device_ref(dev));
+  }
+
+  if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_POINTER))
+  {
+    CLog::Log(LOGDEBUG, "CLibInputHandler::%s - pointer type device added: %s (%s)", __FUNCTION__, name, sysname);
+    m_devices.push_back(libinput_device_ref(dev));
+  }
+
+  if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_KEYBOARD))
+  {
+    CLog::Log(LOGDEBUG, "CLibInputHandler::%s - keyboard type device added: %s (%s)", __FUNCTION__, name, sysname);
+    m_devices.push_back(libinput_device_ref(dev));
+    m_keyboard->GetRepeat(dev);
+  }
+}
+
+void CLibInputHandler::DeviceRemoved(libinput_device *dev)
+{
+  const char *sysname = libinput_device_get_sysname(dev);
+  const char *name = libinput_device_get_name(dev);
+
+  if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_TOUCH))
+  {
+    CLog::Log(LOGDEBUG, "CLibInputHandler::%s - touch type device removed: %s (%s)", __FUNCTION__, name, sysname);
+    auto device = std::find(m_devices.begin(), m_devices.end(), libinput_device_unref(dev));
+    m_devices.erase(device);
+  }
+
+  if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_POINTER))
+  {
+    CLog::Log(LOGDEBUG, "CLibInputHandler::%s - pointer type device removed: %s (%s)", __FUNCTION__, name, sysname);
+    auto device = std::find(m_devices.begin(), m_devices.end(), libinput_device_unref(dev));
+    m_devices.erase(device);
+  }
+
+  if (libinput_device_has_capability(dev, LIBINPUT_DEVICE_CAP_KEYBOARD))
+  {
+    CLog::Log(LOGDEBUG, "CLibInputHandler::%s - keyboard type device removed: %s (%s)", __FUNCTION__, name, sysname);
+    auto device = std::find(m_devices.begin(), m_devices.end(), libinput_device_unref(dev));
+    m_devices.erase(device);
+  }
+}
```

Comment:
If your IDE doesn't insert newlines automatically, can you check GitHub diffs for this red arrow symbol? This causes headaches when devs come along with an IDE that does enforce newlines at the end of files.

Reply:
thanks, fixed, and I've updated my IDE